### PR TITLE
Separate allocator behavior between user mempool and graph capture via `is_graph_capture` flag

### LINKED ISSUE #175015: User MemPool incorrectly inherits CUDA graph capture behavior in `CUDACachingAllocator`
User MemPool and CUDA graph capture both rely on the same internal state (`captures_underway` in `CUDACachingAllocator`) to manage memory pool routing. However, the allocator currently checks `captures_underway.empty()` to trigger graph capture-specific logic, unintentionally applying these behaviors to both graph capture and user mempool usage. 

As a result, when user mempools are active, two major bugs arise: 

1. the deferral of block freeing, intended only for graph capture, is incorrectly used in user mempools
2. OOM recovery mechanisms are disabled during user mempool contexts.

I reported this problem months ago; see #168137 for details.

## Bug 1: Block-free deferral mechanism incorrectly applied to user pools

In `free()`, when a block has been used on multiple streams:

```cpp
// c10/cuda/CUDACachingAllocator.cpp, free()
if (!block->stream_uses.empty()) {
    if (C10_UNLIKELY(!captures_underway.empty())) {
        // Defer the free -- meant for graph capture only
        deferred_blocks.emplace(block, ...);
    } else {
        insert_events(block);  // normal path
    }
}
```

And in `malloc()`, `process_events()` is skipped:

```cpp
// c10/cuda/CUDACachingAllocator.cpp, malloc()
if (C10_LIKELY(captures_underway.empty())) {
    process_events(context);  // queries cudaEvents to recycle blocks
} else {
    // graph_capture_record_stream_reuse path (if enabled)
    free_safe_blocks_in_capture(context, stream);
}
```

Since `use_mem_pool` pushes onto `captures_underway`, the user-pool path hits the graph-capture branch. Blocks with multi-stream usage are put into `deferred_blocks` instead of having events inserted. And `process_events` is skipped, so even single-stream event-based blocks are not recycled. The `free_safe_blocks_in_capture` path is a no-op for user pools because `cudaStreamGetCaptureInfo` returns `cudaStreamCaptureStatusNone`.

**Impact:** Under a user mempool, any block that has been `recordStream`'d to a second stream becomes unreclaimable for the entire duration of the `use_mem_pool` context. The memory is held in `deferred_blocks` or the event queue with no mechanism to process it, increasing memory pressure.

## Bug 2: OOM recovery blocked under user pools

The allocator's last-resort retry in `malloc()`:

```cpp
// c10/cuda/CUDACachingAllocator.cpp, malloc()
block_found = alloc_block(params, false, context, lock)
    || try_mempool_fallback(...)
    || (release_available_cached_blocks(params, context) &&
        alloc_block(params, false, context, lock))
    // Last resort: free ALL cached blocks and retry
    || (C10_LIKELY(captures_underway.empty()) &&
        release_cached_blocks(context, {0, 0}) &&
        alloc_block(params, true, context, lock));
```

The final fallback is gated on `captures_underway.empty()`. During graph capture this is correct, since `release_cached_blocks` calls `cudaEventSynchronize`, which are illegal during capture. But for user mempools, no capture is happening, and these APIs are perfectly safe.

**Impact:** When a user mempool accumulates cached blocks (from prior allocations that were freed back to the caching allocator) and a new allocation request exceeds the largest individual cached block, the allocator cannot reclaim its own cached memory. It fails with OOM even though the device has plenty of reclaimable memory.

This is the root cause of a real-world issue in SGLang reported by @nvcastet here https://github.com/sgl-project/sglang/pull/17089.

## Reproducer

I asked Claude to generate this reproducer. Both tests FAIL on current PyTorch:

<details>
<summary>test_mempool_bugs.py (click to expand)</summary>

```python
import sys
import torch

PASS = 0
FAIL = 0

def report(name, passed, detail=""):
    global PASS, FAIL
    if passed:
        PASS += 1
    else:
        FAIL += 1
    print(f"[{'PASS' if passed else 'FAIL'}] {name}")
    if detail:
        print(f"       {detail}")


def test_defer_mechanism():
    """Bug 1: multi-stream block frees are deferred under user mempool.

    During graph capture, the allocator defers multi-stream block frees
    because events cannot be meaningfully queried (kernels are recorded,
    not executed). For user pools, kernels execute normally, so the block
    should be recyclable after synchronization."""
    torch.cuda.empty_cache()
    pool = torch.cuda.MemPool()
    s2 = torch.cuda.Stream()

    with torch.cuda.use_mem_pool(pool):
        a = torch.empty(1024 * 1024, dtype=torch.uint8, device="cuda")
        ptr = a.data_ptr()

        with torch.cuda.stream(s2):
            a.record_stream(s2)
            _ = a + 1

        del a
        torch.cuda.synchronize()

        b = torch.empty(1024 * 1024, dtype=torch.uint8, device="cuda")
        reused = b.data_ptr() == ptr
        del b

    report(
        "Block free not deferred under user mempool",
        reused,
        "Block was NOT recycled after sync. The free was deferred into "
        "deferred_blocks and process_events was skipped, as if under "
        "graph capture." if not reused else "",
    )
    
def test_oom_recovery():
    """Bug 2: OOM recovery skipped under user mempool.

    Mirrors the real-world reproducer: allocate in a doubling pattern
    (1 GiB, 2 GiB, 4 GiB, ...) so each freed allocation caches a block
    too small for the next request, forcing a fresh cudaMalloc each time.
    Eventually the GPU runs out of fresh memory.  The default allocator
    recovers via release_cached_blocks (frees ALL cached blocks and
    retries).  Under use_mem_pool this path is gated on
    captures_underway.empty() and is skipped, so the allocation fails
    even though the device has plenty of reclaimable cached memory."""
    device = torch.device("cuda:0")
    pool = torch.cuda.MemPool()
    gib = 1024 ** 3

    def max_doubling_alloc(use_pool):
        torch.cuda.empty_cache()
        max_size = 0
        size = 1
        try:
            while size < 2000:
                if use_pool:
                    with torch.cuda.use_mem_pool(pool):
                        torch.empty(gib * size, dtype=torch.uint8, device=device)
                else:
                    torch.empty(gib * size, dtype=torch.uint8, device=device)
                max_size = size
                size *= 2
        except Exception:
            pass
        torch.cuda.empty_cache()
        return max_size

    max_default = max_doubling_alloc(use_pool=False)
    max_pool = max_doubling_alloc(use_pool=True)

    report(
        "OOM recovery works under user mempool",
        max_pool >= max_default,
        f"Default allocator max: {max_default} GiB, "
        f"MemPool max: {max_pool} GiB. "
        f"MemPool hit OOM because release_cached_blocks was skipped "
        f"(gated on captures_underway.empty())."
        if max_pool < max_default else "",
    )



def main():
    if not torch.cuda.is_available():
        print("CUDA not available.")
        sys.exit(1)
    print(f"PyTorch {torch.__version__}")
    print(f"CUDA device: {torch.cuda.get_device_name(0)}\n")
    test_defer_mechanism()
    test_oom_recovery()
    print(f"\nResults: {PASS} passed, {FAIL} failed")
    sys.exit(1 if FAIL else 0)

if __name__ == "__main__":
    main()
```

</details>

## Proposed Fix

Rather than that full refactoring in #168137, We can targeted minimal fix as what @ngimel [suggested](https://github.com/pytorch/pytorch/pull/168137#issuecomment-3554136351), keeping the existing structures and handling operations differently depending on configuration.

Add a `bool is_graph_capture` flag to each entry in the `captures_underway` vector (via `beginAllocateToPool`). `CUDAGraph::capture_begin` would pass `is_graph_capture=true`; `use_mem_pool` would pass `false`. A new predicate `graph_capture_underway()` would return `true` only when any entry has the flag set. The safety guards (deferring frees, skipping `process_events`, blocking `release_cached_blocks`) would switch from `captures_underway.empty()` to `graph_capture_underway()`, so they only activate during actual graph capture. Pool routing continues to use `captures_underway` as before, both graph capture and user pools still need it.

Separately, @galv [raised](https://github.com/pytorch/pytorch/pull/168137#discussion_r2540626505) the idea of directly querying `cudaStreamIsCapturing()` at allocation/free time to distinguish capture state, rather than relying on internal bookkeeping. That could be another alternative.

@eqy @syed-ahmed @galv @ngimel

cc @ptrblck @msaroufim @eqy @jerryzh168 @tinglvv @nWEIdia @mcarilli @ezyang @eellison @penguinwu @BoyuanFeng

### PR DESCRIPTION
**Pull request by eee4017**
Fixes #175015

Resubmits #168137

Rebased onto #181830 

```cpp
bool is_capture_context() const {
  if (C10_LIKELY(graph_capture_count_.load(std::memory_order_relaxed) == 0)) {
    return false;
  }
  cudaStream_t stream = cuda::getCurrentCUDAStream(device_id).stream();
  if (stream_get_capture_info(stream).status != cudaStreamCaptureStatusNone) {
    return true;
  }
  return false;
}
```

cc @ptrblck @msaroufim @eqy @jerryzh168 @tinglvv @nWEIdia @mcarilli @ezyang @eellison @penguinwu @BoyuanFeng @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by eqy** [comment]
needs cleanup o7

**Review by jansel — CHANGES_REQUESTED** [review]
fix merge conflicts

**Comment by ngimel** [comment]
This PR would require manual adjustment, don't let claude do it

**Comment by jansel** [comment]
@ngimel or others should handle the review here

**Review by eqy — APPROVED** [review]

**Review by ngimel — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Comment by anijain2305** [comment]
Claude attributes this PR to be the suspect of "torchbench cudagraphs inference exec time +9%"

  3. Execution Time (torchbench cudagraphs inference): REAL regression, suspect identified

  5-day trend:
  - May 10-14: stable at ~35s
  - May 15: jumps to 38.6s (+9%)

  Prime suspect: c1d7692bb69 — Separate allocator behavior for graph capture (#175374)

  This commit changes the CUDA caching allocator to distinguish between actual graph capture (where cudaEventQuery is illegal) and user mempool diversions. It replaces fast
  num_active_captures_ == 0 integer checks with is_capture_context() calls that can invoke cudaStreamGetCaptureInfo() — a CUDA driver call. While there's a fast-path early-out when no
  capture is active, the behavior change during/after graph warmup could alter allocation patterns, memory reuse, and event processing timing.

  The cudagraphs_freezing variant also regressed (32.8s -> 35.9s, +9.4%), reinforcing that this is a CUDA graph-specific issue.

**Comment by eee4017** [comment]
@anijain2305 do you know which case has the performance regression? I can take a look.

**Comment by ngimel** [comment]
Can you confirm by running with this diff backed out? If so we can revert

**Comment by anijain2305** [comment]
I was led to a wrong path from Claude analysis. Confirmed that this is not a regression, everything is well within noise

This is what I ran manually (for a few more models)

```

  cd /data/users/anijain/pytorch3

  # --- BEFORE ---
  git checkout c1d7692bb69f^ --quiet
  git submodule sync --quiet && git submodule update --init --recursive --quiet
  MAX_JOBS=32 python3 setup.py develop

  python3 benchmarks/dynamo/torchbench.py --inductor --performance --inference --bfloat16 -dcuda --only resnet18 -n3 --output before_resnet18_inductor.csv
  python3 benchmarks/dynamo/torchbench.py --backend cudagraphs --performance --inference --bfloat16 -dcuda --only resnet18 -n3 --output before_resnet18_cudagraphs.csv

  # --- AFTER ---
  git checkout c1d7692bb69f --quiet
  git submodule sync --quiet && git submodule update --init --recursive --quiet
  MAX_JOBS=32 python3 setup.py develop

  python3 benchmarks/dynamo/torchbench.py --inductor --performance --inference --bfloat16 -dcuda --only resnet18 -n3 --output after_resnet18_inductor.csv
  python3 benchmarks/dynamo/torchbench.py --backend cudagraphs --performance --inference --bfloat16 -dcuda --only resnet18 -n3 --output after_resnet18_cudagraphs.csv

```

and there is no regression.
